src/wasm/artifact_cache.rs

```diff
@@ -0,0 +1,57 @@
+use super::fingerprint::Fingerprint;
+use crate::Result;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    path::{Path, PathBuf},
+};
+
+#[derive(Derivative, Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+#[repr(C)]
+pub struct ArtifactCache {
+    fingerprints: HashMap<PathBuf, Fingerprint>,
+    root: PathBuf,
+}
+
+impl ArtifactCache {
+    pub fn new(root: impl Into<PathBuf>) -> Result<Self> {
+        let root = root.into();
+
+        match std::fs::File::open(root.clone().join(".fingerprints")) {
+            Ok(fingerprint_file) => {
+                let reader = std::io::BufReader::new(fingerprint_file);
+                let fingerprints = serde_json::from_reader(reader)?;
+                Ok(Self { fingerprints, root })
+            }
+            Err(e) if e.kind() == std::io::ErrorKind::NotFound => Ok(Self {
+                fingerprints: HashMap::new(),
+                root,
+            }),
+            Err(e) => Err(Box::new(e)),
+        }
+    }
+
+    /// Returns true if the file is fresh and needs to be rebuilt.
```

Comment:
Is this correct? Seems like it would not need to be rebuilt if it matches.

Reply:
Docs are wrong. :(